pandas/core/groupby/generic.py

```diff
@@ -622,33 +625,27 @@ def value_counts(
         ids, val = ids[mask], val[mask]
 
         if bins is None:
-            lab, lev = algorithms.factorize(val, sort=True)
+            lab, lev = cast(
+                Tuple[np.ndarray, Index], algorithms.factorize(val, sort=True)
+            )
             llab = lambda lab, inc: lab[inc]
         else:
 
-            # lab is a Categorical with categories an IntervalIndex
-            lab = cut(Series(val), bins, include_lowest=True)
-            # error: "ndarray" has no attribute "cat"
-            lev = lab.cat.categories  # type: ignore[attr-defined]
-            # error: No overload variant of "take" of "_ArrayOrScalarCommon" matches
-            # argument types "Any", "bool", "Union[Any, float]"
-            lab = lev.take(  # type: ignore[call-overload]
-                # error: "ndarray" has no attribute "cat"
-                lab.cat.codes,  # type: ignore[attr-defined]
+            # labels is a Series with categories an IntervalIndex
+            labels: Series = cut(Series(val), bins, include_lowest=True)
+            lev = labels.cat.categories
+            lab = lev.take(
+                labels.cat.codes,
                 allow_fill=True,
-                # error: Item "ndarray" of "Union[ndarray, Index]" has no attribute
-                # "_na_value"
-                fill_value=lev._na_value,  # type: ignore[union-attr]
+                fill_value=lev._na_value,
             )
             llab = lambda lab, inc: lab[inc]._multiindex.codes[-1]
 
         if is_interval_dtype(lab.dtype):
             # TODO: should we do this inside II?
 
-            # error: "ndarray" has no attribute "left"
-            # error: "ndarray" has no attribute "right"
             sorter = np.lexsort(
-                (lab.left, lab.right, ids)  # type: ignore[attr-defined]
+                (cast(Interval, lab).left, cast(Interval, lab).right, ids)
```

Comment:
@simonjayhawkins might be a better person to judge when/where to use casts. I think I also sometime use them where I shouldn't use them :)

Reply:
A cast here is probably OK since mypy can't narrow from `is_interval_dtype`

I think I've look at using TypeGuard for some of the these functions, but unfortunately we allow `arr_or_dtype` as the argument so we would end up returning a union return type and so does not solve the issue in many cases. For TypeGuard to be effective we would have to limit the input to the is_* functions.

For clarity, I personally would put the cast directly adjacent to the `is_interval_dtype` condition. (in general I prefer the cast to be a single statement rather than inlining in other expressions)